### DISCUSSION TITLE
Show & Tell: WorldModel architecture — predictive dynamics for model-based planning

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#369] [Show and tell]
## WorldModel — architecture deep-dive

Issue #368 introduces `WorldModel`, the Phase 13.1 component that gives ASI-Build the ability to simulate "what happens next" before committing to an action.

---

### Component map

```
CognitiveCycle
│
├── _model_based_step(obs, candidate_actions)
│     └── WorldModel.dream_rollout(obs, action_seq)
│           └── _forward(ModelInput) × rollout_horizon
│                 └── ModelOutput(next_obs, reward, done, error)
│
└── _update_world_model(inp, actual_next_obs, reward, done)
      └── WorldModel.update() → SGD loss
            └── deque replay buffer (max 100k)
```

### TransitionBackend selection guide

| Backend | Latency | When to use |
|---------|---------|-------------|
| `MLP` | ~0.1ms | Short-horizon, fast iteration |
| `LSTM` | ~0.5ms | Temporal dependencies |
| `TRANSFORMER` | ~2ms | Long observation histories |
| `ENSEMBLE` | ~N×MLP | Exploration + epistemic uncertainty |

### Dream rollout — surprise detection

Surprise score = mean MSE(predicted_obs, actual_obs) across rollout steps.
When `surprise_score > surprise_threshold`, CognitiveCycle triggers re-planning.

### WorldModelConfig tuning

| Parameter | Default | Lower → | Higher → |
|-----------|---------|---------|----------|
| `rollout_horizon` | 10 | Faster/myopic | Slower/far-sighted |
| `surprise_threshold` | 0.05 | More re-plans | Fewer re-plans |
| `ensemble_size` | 5 | Less uncertainty | More precision |
| `learning_rate` | 3e-4 | Slow adapt | Instability risk |
| `max_buffer_size` | 100k | Less memory | More off-policy data |

### 5 Prometheus metrics

| Metric | Type | Alert |
|--------|------|-------|
| `wm_update_total` | Counter | <10/min → not learning |
| `wm_train_loss` | Gauge | >0.1 → diverging |
| `wm_surprise_score` | Gauge | >0.05 → distribution shift |
| `wm_buffer_size` | Gauge | <1000 → insufficient data |
| `wm_rollout_steps_total` | Counter | rate drop → planner idle |

### Open questions

1. **Latent space vs raw obs** — predict in raw obs space or learned latent (Dreamer/RSSM)? Latent is more scalable but adds encoder/decoder.
2. **Ensemble disagreement as exploration bonus** — add ENSEMBLE variance to reward signal?
3. **Offline pre-training** — batch pre-train before online loop, or cold-start always?

Related: Issue #368 | Q&A #370 | Wiki: Phase-13-World-Model (coming soon)
